plutus-tutorial/doc/plutus-tx.adoc

```diff
@@ -0,0 +1,383 @@
+    Fixed n -> n < current
+    Never -> False
+   ||])
+----
+
+== The Plutus Tx Prelude
+
+The `Language.PlutusTx.Prelude` module is a drop-in replacement for the
+normal Haskell Prelude, but with some functions and typeclasses redefined to be easier
+for the Plutus Tx compiler to handle (i.e. `INLINABLE`).
+
+You should use the Plutus Tx Prelude whenever you are writing code that you expect to compile with
+the Plutus Tx compiler. All of the definitions in the Plutus Tx Prelude have
+working Haskell definitions, so you can use them in normal Haskell code too,
+although the Haskell Prelude versions will probably perform better.
+
+To use the Plutus Tx Prelude, use the `NoImplicitPrelude` language
+pragma, and import `Language.PlutusTx.Prelude`.
+
+Plutus Tx has some builtin types and functions available for working
+with primitive data (integers and bytestrings), as well as a few special
+functions. These builtins are also exported from the Plutus Tx Prelude.
+
+The `error` builtin deserves a special mention. `error` causes the
+transaction to abort when it is evaluated, which one way to trigger
+validation failure.
+
+== Lifting values
+
+So far we’ve seen how to define pieces of code _statically_ (when you
+_compile_ your main Haskell program), but you are likely to want to generate code
+_dynamically_ (when you _run_ your main Haskell program). For example, you
+might be writing the body of a transaction to initiate a crowdfunding
+smart contract, which would need to be parameterized by data
+determining the size of the goal, the campaign start and end times, etc.
+
+We can do this in the same way that we normally parameterize code in functional
+programming: we write the static code as a _function_, and we provide the
+argument later to configure it.
+
+In our case we have a complication, in that we want to make the argument and
+apply the function to it at runtime. Plutus Tx provides a mechanism to do this
+called _lifting_. Lifting makes it easy to use the same types
+both inside your Plutus Tx program and in the external code that uses
+it.
+
+As a very simple example, let’s write an add-one
+function.
+
+[source,haskell]
+----
+addOne :: CompiledCode (Integer -> Integer)
+addOne = $$(compile [|| \(x:: Integer) -> x `addInteger` 1 ||])
+----
+
+Now, suppose we want to apply this to `4` at runtime, giving us a
+program that computes to `5`. We need to _lift_ the argument (`4`)
+from Haskell to Plutus Core, and then we need to apply the function to
+it.
+
+[source,haskell]
+----
+addOneToN :: Integer -> CompiledCode Integer
+addOneToN n =
+    addOne
+    `applyCode` -- <.>
+    unsafeLiftCode n -- <.>
+
+{- |
+>>> let program = getPlc $ addOneToN 4
+>>> pretty program
+(program 1.0.0
+  [
+    [
+      (lam
+        addInteger
+        (fun (con integer) (fun (con integer) (con integer)))
+        (lam ds (con integer) [ [ addInteger ds ] (con 1) ])
+      )
+      (builtin addInteger)
+    ]
+    (con 4)
```

Comment:
Maybe point out that the argument 4 now shows up in the PLC (Not sure if you can do this without messing up the doctests)

Reply:
More generally, I think this is a good example but the effect of `applyCode` would be easier to understand if there was a comparison of the PLC of `addToOne` and that of `addToOneN 4`.

Reply:
Yeah, I can't put a callout in the doctest output, unfortunately...